File: seqsvr/base/set.cc
/*
 *  Copyright (c) 2016, https://github.com/nebula-im
 *  All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "base/set.h"

#include "nebula/base/logger/glog_util.h"

uint64_t Set::SetMaxSeq(uint32_t id, uint64_t max_seq) {
  auto idx = CalcSectionID(set_id, id);
  if (!idx.first) {
    LOG(ERROR) << "SetSectionsData - max_seq invalid: local seq = "
      << id << ", req seq = " << max_seq
      << ", in set: " << ToString();
    
    return 0;
  }
  
  if (max_seq > set_max_seqs_data[idx.second]) {
    max_seq = (max_seq/kSeqStep+1)*kSeqStep;
    set_max_seqs_data[idx.second] = max_seq;
  }
  
  return max_seq;
}

File: proto/seqsvr/set.go
// Copyright 2022 Teamgram Authors
//  All rights reserved.
//
// Author: Benqi (wubenqi@gmail.com)
//

package seqsvr

import (
	"encoding/binary"

	"github.com/edsrzf/mmap-go"
)

type SliceInt64 struct {
	mmap.MMap
}

func (s SliceInt64) SetIdx(idx int, v int64) {
	binary.LittleEndian.PutUint64(s.MMap[idx<<3:], uint64(v))
	s.Flush()
}

func (s SliceInt64) GetIdx(idx int) int64 {
	return int64(binary.LittleEndian.Uint64(s.MMap[idx<<3:]))
}

func (s SliceInt64) Len() int {
	return len(s.MMap) / 8
}

type Set struct {
	SetId          SetID
	SetMaxSeqsData SliceInt64
}

func (s *Set) ID() SetID {
	return s.SetId
}

func (s *Set) SetMaxSeq(id int32, maxSeq int64) int64 {
	ok, idx := s.SetId.CalcSectionID(id)
	if !ok {
		//LOG(ERROR) << "SetSectionsData - max_seq invalid: local seq = "
		//<< id << ", req seq = " << max_seq
		//<< ", in set: " << ToString();
		return 0
	}

	v := s.SetMaxSeqsData.GetIdx(int(idx))
	if maxSeq > v {
		maxSeq = (maxSeq/SeqStep + 1) * SeqStep
		s.SetMaxSeqsData.SetIdx(int(idx), maxSeq)
	}

	return maxSeq
}
